File: 15.BinarySearch二分查找/leetcode_1871_canReach.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    bool canReach(string s, int minJump, int maxJump) {
        int n = s.size();
        if (s[n-1] == '1'){
            return false;
        }
        vector<int> pos;
        for (int i=0; i<n; i++){
            if (s[i] == '0'){
                pos.push_back(i);
            }
        }
        int begin = 0;
        vector<int> q = {0};
        while (!q.empty()){
            vector<int> q2;
            for (int i : q){
                if (i == n-1){
                    return true;
                }
                int j = lower_bound(pos.begin(), pos.end(), i+minJump)-pos.begin();
                j = max(j, begin);
                while (j<pos.size() && pos[j]-i >= minJump && pos[j]-i <= maxJump){
                    q2.push_back(pos[j]);
                    j++;
                }
                begin = j;
            }
            q = q2;
        }
        return false;
    }
};

int main(){
    Solution sl;
    string s = "011010";
    int minJump = 2, maxJump = 3;
    cout << sl.canReach(s, minJump, maxJump) << endl;
}

File: 07.BFS广度优先搜索/leetcode_1871_canReach.py
from bisect import *

class Solution:
    def canReach(self, s: str, minJump: int, maxJump: int) -> bool:
        if s[-1] == '1':
            return False
        
        n = len(s)
        q = [0]
        pos = [i for i in range(1, n) if s[i]=='0']
        begin = 0
        while q:
            q2 = []
            for i in q:
                if i == n-1:
                    return True
                j = max(begin, bisect_left(pos, i+minJump))
                while j<len(pos) and pos[j]-i >= minJump and pos[j]-i <= maxJump:
                    q2.append(pos[j])
                    j += 1
                begin = j
            q = q2
        return False

if __name__ == "__main__":
    sl = Solution()
    s = "011010"
    minJump, maxJump = 2, 3
    print(sl.canReach(s, minJump, maxJump))
